phoenix-core/src/main/java/org/apache/phoenix/execute/MutationState.java

```diff
@@ -488,7 +490,55 @@ public void rollback(PhoenixConnection connection) throws SQLException {
         numRows = 0;
     }
     
+    private Set<Integer> getOrderOfUncommittedStatements() {
+    	Set<Integer> result = newHashSet();
+    	for (Map<ImmutableBytesPtr, RowMutationState> rowMutations : mutations.values()) {
+    		for (RowMutationState rowMutationState : rowMutations.values()) {
+    			result.addAll(rowMutationState.getOrderOfStatementsInConnection());
+    		}
+    	}
+    	return result;
+    }
+    
     @Override
     public void close() throws SQLException {
     }
+    
+    public static class RowMutationState {
+    	private Map<PColumn,byte[]> columnValues;
+    	private Set<Integer> orderOfStatementsInConnection;
```

Comment:
I'm somewhat worried about the memory bloat of adding another Set object here. An alternative would be a long[] that acts as a bit set where any set bit represents the statement index. It's unlikely that there'd be more than 64 statements in a commit, but if there were, you could always re-allocate the long[]. You could then materialize the set of statement indexes from it. 

Another more flexible alternative would be to make RowMutationState an interface and have a SingleStatementRowMutationState that has a single int index member variable (likely the common case). Then, add a join method that returns a new MultiRowMutationState with a long as a bitset and maybe one more with a long[] as a bitset.